crates/op-consensus/src/receipt/envelope.rs

```diff
@@ -0,0 +1,216 @@
+use crate::{OpReceipt, OpReceiptWithBloom, OpTxType};
+use alloy_eips::eip2718::{Decodable2718, Encodable2718};
+use alloy_primitives::{Bloom, Log};
+use alloy_rlp::{length_of_length, BufMut, Decodable, Encodable};
+
+/// Receipt envelope, as defined in [EIP-2718], modified for OP Stack chains.
+///
+/// This enum distinguishes between tagged and untagged legacy receipts, as the
+/// in-protocol merkle tree may commit to EITHER 0-prefixed or raw. Therefore
+/// we must ensure that encoding returns the precise byte-array that was
+/// decoded, preserving the presence or absence of the `TransactionType` flag.
+///
+/// Transaction receipt payloads are specified in their respective EIPs.
+///
+/// [EIP-2718]: https://eips.ethereum.org/EIPS/eip-2718
+#[derive(Debug, Clone, PartialEq, Eq)]
+#[cfg_attr(feature = "serde", derive(serde::Serialize, serde::Deserialize))]
+#[cfg_attr(feature = "serde", serde(tag = "type"))]
+#[non_exhaustive]
+pub enum OpReceiptEnvelope<T = Log> {
+    /// Receipt envelope with no type flag.
+    #[cfg_attr(feature = "serde", serde(rename = "0x0", alias = "0x00"))]
+    Legacy(OpReceiptWithBloom<T>),
+    /// Receipt envelope with type flag 1, containing a [EIP-2930] receipt.
+    ///
+    /// [EIP-2930]: https://eips.ethereum.org/EIPS/eip-2930
+    #[cfg_attr(feature = "serde", serde(rename = "0x1", alias = "0x01"))]
+    Eip2930(OpReceiptWithBloom<T>),
+    /// Receipt envelope with type flag 2, containing a [EIP-1559] receipt.
+    ///
+    /// [EIP-1559]: https://eips.ethereum.org/EIPS/eip-1559
+    #[cfg_attr(feature = "serde", serde(rename = "0x2", alias = "0x02"))]
+    Eip1559(OpReceiptWithBloom<T>),
+    /// Receipt envelope with type flag 3, containing a [EIP-4844] receipt.
+    ///
+    /// [EIP-4844]: https://eips.ethereum.org/EIPS/eip-4844
+    #[cfg_attr(feature = "serde", serde(rename = "0x3", alias = "0x03"))]
+    Eip4844(OpReceiptWithBloom<T>),
+    /// Receipt envelope with type flag 126, containing a [deposit] receipt.
+    ///
+    /// [deposit]: https://specs.optimism.io/protocol/deposits.html
+    #[cfg_attr(feature = "serde", serde(rename = "0x7E", alias = "0x7E"))]
+    Deposit(OpReceiptWithBloom<T>),
```

Comment:
hmm, only this variant needs the Op specific receipt type, right?

the others could reuse the the alloy-consensus one?

Reply:
Yeah, only this variant needs it. Decided to use all `OpReceiptWithBloom`s to avoid type conversion within `as_receipt_with_bloom` + `as_receipt`.

Technically, all of the receipts within OP consensus do use this format, though the extra `deposit_nonce` + `deposit_receipt_version` fields are omitted from the RLP if they don't exist. Could reuse the `ReceiptWithBloom<T>` type for others -- would you rather have the type conversions?